crates/storage/src/connection/event.rs

```diff
@@ -220,6 +218,101 @@ impl Transaction<'_> {
         }
     }
 
+    #[cfg(feature = "aggregate_bloom")]
+    pub fn events_in_range_aggregate(
+        &self,
+        from_block: BlockNumber,
+        to_block: BlockNumber,
+        contract_address: Option<ContractAddress>,
+        keys: Vec<Vec<EventKey>>,
+    ) -> anyhow::Result<(Vec<EmittedEvent>, Option<BlockNumber>)> {
+        let Some(latest_block) = self.block_number(crate::BlockId::Latest)? else {
+            // No blocks in the database
+            return Ok((vec![], None));
+        };
+        if from_block > latest_block {
+            return Ok((vec![], None));
+        }
+
+        let filter = EventFilter {
+            contract_address,
+            keys,
+            page_size: usize::MAX - 1,
+            ..Default::default()
+        };
+
+        let aggregates = self.load_aggregate_bloom_range(from_block, to_block)?;
+
+        let blocks_to_scan = aggregates
+            .iter()
+            .flat_map(|aggregate| aggregate.blocks_for_filter(&filter))
+            .filter(|&block| (from_block..=to_block).contains(&block));
+
+        let key_filter_is_empty = filter.keys.iter().flatten().count() == 0;
+        let keys: Vec<std::collections::HashSet<_>> = filter
+            .keys
+            .iter()
+            .map(|keys| keys.iter().collect())
+            .collect();
+
+        let mut emitted_events = vec![];
+
+        for block in blocks_to_scan {
+            let Some(block_header) = self.block_header(crate::BlockId::Number(block))? else {
+                break;
+            };
+
+            let events = match self.events_for_block(block.into())? {
+                Some(events) => events,
+                // Reached the end of P2P (checkpoint) synced events.
+                None => break,
+            };
```

Comment:
BTW: IMO we shouldn't be serving JSON-RPC requests while still at the checkpoint sync stage. There's too many things that might be just missing from the DB and most of our RPC code will just emit internal errors for those cases.

Reply:
Absolutely, it has nothing to do with this PR -- it's an unrelated deficiency. I just though I'd mention that since you've been adding some comments documenting behavior in that specific case.